### DISCUSSION TITLE
Drawing a Node Graph?

### DISCUSSION THREAD

**Discussion by sdfgeoff** [emilk/egui#659] [Q&A]
First up, wow. egui is great. I've never put together a UI so fast in my life!

Currently I have a project like this:
![egui](https://user-images.githubusercontent.com/13490050/131124548-d3b6ddc2-a1aa-486d-a807-9171a4f9458c.jpg)

My aim is to make something like this:
![what-i-want-to-do](https://user-images.githubusercontent.com/13490050/131124572-115014d0-2e6e-4544-94f5-f5f60e9075d1.jpg)

At the moment my main question is drawing - how can I render lines that don't "belong" to a widget? I'm fine with drawing on the background element, but then I need to get the floating windows positions somehow? I have a depressing feeling this may not be architecturally possible with an immediate mode rendering library without pulling the state data out of it's innards.

The exact form of the lines doesn't concern me - straight lines are fine. Hacking around inside egui is also fine - but I just don't know where to start.

**Comment by coderedart** [comment]
not an answer, but you could look at https://github.com/haighcam/egui_nodes if that helps

**Reply by sdfgeoff** [reply]
Oh wow, perfect. Looks like it can do exactly what I want!

**Comment by emilk** [comment]
You can use https://docs.rs/egui/0.14.0/egui/struct.CtxRef.html#method.layer_painter to get a painter (use something like `LayerId::new(Order::Foreground, Id::new("node_graph_lines"))`).

To know where to draw, I suggest you store away all the connection endpoints somewhere while you are drawing your node windows, and then draw the connections last.
